### ISSUE TITLE
Remove JoinableQueue again.

### ISSUE THREAD

**Issue by ajdavis** [python/asyncio#234]
In Python 3.5, we removed the deprecated asyncio.queues.JoinableQueue:

http://bugs.python.org/issue23464

Let's remove the deprecated alias from this repository as well.

Also, can I get some guidance on the Tulip development process? Should all changes happen here, or some changes happen in the CPython repository, or what? Where should issues be opened?

**Comment by gvanrossum** [comment]
I'm not sure what to do with this. This is the first time that we introduced something in the 3.5 version of asyncio that we didn't also introduce in 3.4.

The workflow has been various -- people have filed bugs and feature requests in the Python tracker as well as in the asyncio tracker, and different developers have made changes in the CPython repo and in the asyncio repo. Copying files back and forth is supported by the "update_stdlib.sh" script in the asyncio repo, which shows files that differ between the two repos and offers to copy files from the asyncio to the CPython repo ("y") or in the opposite direction ("back"). But you have to manually commit and push the changes in the destination repo, and when copying to the CPython repo you have to copy it both to the default branch and the 3.4 branch (and I think technically you should copy to the 3.4 branch and then merge forward). Backward incompatible changes in the 3.4 branch are allowed because of the provisional status. (But not that the selectors.py module is not provisional and changes to it should not be copied to the 3.4 branch -- the update script handles this.)

Maybe we should just keep changing the asyncio library in successive 3.4.x releases (as we've done), and damn the torpedoes (incompatibilities). Since asyncio is provisional in 3.4, I think this is technically okay -- PEP 411 does not give very precise guidance on this.

My own preferred workflow (especially for features) would then be to build a feature in the asyncio repo, do some testing and reviewing by others using commits in that repo for exchange, then once it's deemed stable copy it to the 3.4 branch using the update script, add the docs, commit, and then merge forward. Note that there are no docs in the asyncio repo (except for docstrings), and all docs originate in the 3.4 doc tree.

Once 3.5 is out we either have to freeze the package or do feature development in the 3.6 branch only, and the asyncio repo is then effectively frozen (except for bugfixes, which should be copied to 3.4, 3.5, 3.6, and the asyncio repo).

OK, so now I propose to delete the JoinableQueue alias from 3.4 as well, preserving the "one source to bind them all" philosophy at least for now. Thoughts?

**Comment by gvanrossum** [comment]
So actually I think in this particular case I'd like to keep the `JoinableQueue` alias in the 3.4 branch until one release has gone out with that alias. In 3.4.3 there was still a separate `JoinableQueue` queue, so I think 3.4.4 should go out with the alias, and in 3.4.5 it can be gone. Until then we should keep it here in the asyncio repo too.

**Comment by ajdavis** [comment]
Thanks Guido, makes sense to me. Would you like to add me as a collaborator again, so I can assign myself this issue?

**Comment by ajdavis** [comment]
Thanks for adding me as a collaborator. I signed the PSF contributor agreement March 18, 2013 via EchoSign.

**Comment by gvanrossum** [comment]
I'd like to keep this issue open until we've also removed JoinableQueue from the 3.4 tree and from the asyncio repo. Right now it's only been removed from the 3.5 (default) branch of CPython. We can remove it from 3.4 after the next 3.4 bugfix release.

**Comment by haypo** [comment]
I propose to to use conditional code to keep same code in github asyncio, asyncio of python 3.4 and asyncio of python 3.5.

```
if sys.version_info < (3, 5):
    JoinableQueue = Queue
    """Deprecated alias for Queue."""
```

With the if, the code can be the same everywhere, and the symbol is only available in Python 3.3 and 3.4.

> We can remove it from 3.4 after the next 3.4 bugfix release.

Why do you want to remove JoinableQueue from Python 3.4? I will be annoying have to check the Python minor version to use Queue or JoinableQueue. It's already annoying that create_task() is only available since Python 3.4.2.

**Comment by gvanrossum** [comment]
I don't care either way -- PEP 3056 was provisional in 3.4.

**Comment by haypo** [comment]
Fixed by commit 1f85dc7d6ba07d1a8be18446d251ebe5b7575d34.

asynico.JoinableQueue symbol is no more available on Python 3.5 and newer.

**Comment by ajdavis** [comment]
Great, thanks for putting this to bed.
